### PR TITLE
Add documentation on out-of-band tokens and the /validate/polltransaction endpoint

### LINKED ISSUE #1838: Add an endpoint for polling the transaction status which automatically deletes answered challenges
Assume a user has enrolled a push token and tries to log in via the owncloud app. The app triggers a challenge in the privacyIDEA server, the user smartphone receives a push notification, the user accepts the challenge. The smartphone app sends a challenge response to the ``/ttype/push`` endpoint, which sets the ``otp_valid`` flag of the challenge in the database. According to the [push token protocol](https://github.com/privacyidea/privacyidea/wiki/concept%3A-PushToken#polling-endpoint-for-application), the owncloud app polls the ``/token/challenges`` endpoint and authenticates the user if it finds a challenge with ``otp_valid`` set to true.

After authentication, the accepted challenge *stays* in the database (https://github.com/privacyidea/privacyidea-owncloud-app/issues/73). Eventually, it will be removed by the challenge janitor, but the cleanup timestamp is not really defined: The challenge janitor is, for example, invoked when a new challenge is triggered. But the challenge could potentially stay in the database for a long time.

This means that theoretically, one could still authenticate with the old transaction ID until the challenge is cleaned up. An app should not allow the user to specify a transaction ID, but privacyIDEA should not allow this behavior nonetheless.

An initial idea was to add a boolean flag to the response of ``/token/challenges`` that indicates the validity of the challenge (#1837). However, a better idea might be the following:

We add a new endpoint, like ``POST /token/finalize_transaction`` (or so), which takes a parameter ``transaction_id`` and

1) checks for a valid challenge with the given transaction ID and ``otp_valid=true``
2) if such a challenge exists, the endpoint returns ``true`` and removes *all* challenges with the given transaction ID

This endpoint could be polled by apps to check the challenge response status of out-of-band tokens (currently push + tiqr).

### PR DESCRIPTION
**Pull request by fredreichbier**
Closes #1838

### CODE REVIEW

File: doc/configuration/authentication_modes.rst [hunk 1]
```diff
@@ -0,0 +1,133 @@
...
+``challenge`` and ```outofband``.
+
+Application plugins need to implement the three authentication modes
+separately, as the modes differ in their user experience. For example:
+
+* The HOTP token type implements the ``authenticate`` mode, which is a
+  single-shot authentication flow. For each authentication request, the user
+  uses their token to generate a new HOTP value and enters it along with their
+  OTP PIN. The plugin sends both values to privacyIDEA, which decides whether
+  the authentication is valid or not.
```
Comment (cornelinux) [inline doc/configuration/authentication_modes.rst]: There is only one authentication request

File: doc/configuration/authentication_modes.rst [hunk 2]
```diff
@@ -0,0 +1,133 @@
...
+* The E-Mail token type implements the ``challenge`` mode. With an E-Mail token,
```
Comment (cornelinux) [inline doc/configuration/authentication_modes.rst]: "The E-Mail token or SMS-Token type impement"

File: doc/configuration/authentication_modes.rst [hunk 3]
```diff
@@ -0,0 +1,133 @@
...
+  the authentication flow consists of two steps: In a
+  first step, the plugin triggers a challenge. privacyIDEA sends the challenge
+  response --- a fresh OTP value --- to the user via E-Mail.
+  In a second step, the user responds to the challenge by entering the
+  respective OTP value in the plugin's login form. The plugin sends the
+  challenge response to privacyIDEA, which decides whether the authentication
+  is valid or not.
+* The PUSH token type implements the ``outofband`` mode. With a PUSH token,
```
Comment (cornelinux) [inline doc/configuration/authentication_modes.rst]: "or TiQR token type"

File: doc/configuration/authentication_modes.rst [hunk 4]
```diff
@@ -0,0 +1,133 @@
...
+  the authentication step also consists of two steps:
+  In a first step, the user triggers a challenge. privacyIDEA pushes the
+  challenge to the user's smartphone app. In a second step, the user approves
+  the challenge on their phone, and the app responds to the challenge on behalf
```
Comment (cornelinux) [inline doc/configuration/authentication_modes.rst]: it responds to the challenge by communicating to privacyIDEA server on behalf of the user.

File: doc/configuration/authentication_modes.rst [hunk 5]
```diff
@@ -0,0 +1,133 @@
...
+  and privacyIDEA returns whether the authentication request has succeeded
+  or not.
+
+.. _authentication_mode_challenge:
+
+``challenge`` mode
+------------------
+
+ * The plugin triggers a challenge, for example via the
+   ``/validate/triggerchallenge`` endpoint:
```
Comment (cornelinux) [inline doc/configuration/authentication_modes.rst]: we should either add a reference to challenge response or we should note, that the challenge is also triggered by sending the PIN via /validate/check

File: doc/configuration/authentication_modes.rst [hunk 6]
```diff
@@ -0,0 +1,133 @@
...
+
+   .. code-block:: text
+
+     GET /validate/polltransaction
+
+     transaction_id=<transaction_id>
+
+   If this endpoint returns ``false``, the challenge has not been answered yet.
+ * The user approves the challenge on a separate device, e.g. their
+   smartphone app. The app communicates with a tokentype-specific endpoint
```
Comment (cornelinux) [inline doc/configuration/authentication_modes.rst]: tokentype-specific endpoint of privacyIDEA!

File: doc/configuration/authentication_modes.rst [hunk 7]
```diff
@@ -0,0 +1,133 @@
...
+   finalize the authentication via the ``/validate/check`` endpoint, sending
+   an empty password:
+
+   .. code-block:: text
+
+     POST /validate/check
+
+     user=<user>&transaction_id=<transaction_id>&pass=
+
+   This step is crucial because the ``/validate/check`` endpoint takes defined
```
Comment (cornelinux) [inline doc/configuration/authentication_modes.rst]: note the empty pass.

File: doc/configuration/authentication_modes.rst (hunk 2 above)
Comment (fredreichbier) [reply doc/configuration/authentication_modes.rst]: thanks! fixed in 076624be5af5d799ff9e8cdf6e6d362f1781e495

File: doc/configuration/authentication_modes.rst (hunk 3 above)
Comment (fredreichbier) [reply doc/configuration/authentication_modes.rst]: fixed in 076624be5af5d799ff9e8cdf6e6d362f1781e495

File: doc/configuration/authentication_modes.rst (hunk 4 above)
Comment (fredreichbier) [reply doc/configuration/authentication_modes.rst]: fixed in 076624be5af5d799ff9e8cdf6e6d362f1781e495

File: doc/configuration/authentication_modes.rst (hunk 5 above)
Comment (fredreichbier) [reply doc/configuration/authentication_modes.rst]: fixed in 076624be5af5d799ff9e8cdf6e6d362f1781e495

File: doc/configuration/authentication_modes.rst (hunk 6 above)
Comment (fredreichbier) [reply doc/configuration/authentication_modes.rst]: fixed in 076624be5af5d799ff9e8cdf6e6d362f1781e495

File: doc/configuration/authentication_modes.rst (hunk 7 above)
Comment (fredreichbier) [reply doc/configuration/authentication_modes.rst]: made this more explicit in 076624be5af5d799ff9e8cdf6e6d362f1781e495

**Pull request merged**